spec/models/product_spec.rb

```diff
@@ -0,0 +1,142 @@
+require "rails_helper"
+
+RSpec.describe Product, type: :model do
+  subject {product}
+
+  let!(:supplier) { FactoryBot.create :supplier }
+  let!(:category) { FactoryBot.create :category }
+  let!(:product) {
+    FactoryBot.build :product,
+      supplier_id: supplier.id, category_id: category.id
+  }
+
+  describe "associations" do
+    it "belong to category" do
+      is_expected.to belong_to :category
+    end
+
+    it "belong to supplier" do
+      is_expected.to belong_to :supplier
+    end
+
+    it "has many image" do
+      is_expected.to have_many :images
+    end
+
+    it "has many order_details" do
+      is_expected.to have_many :order_details
+    end
+
+    it "has many comments" do
+      is_expected.to have_many :comments
+    end
+  end
+
+  describe "Delegate" do
+    it { is_expected.to delegate_method(:name).to(:category).with_prefix(true) }
+  end
+
+  describe "Validations" do
+    before {product.save}
```

Comment:
có những loại before nào, khác gì nhau

Reply:
before để thực hiện một đoạn codes nào đó trước khi chạy một example:
Có 2 loại before. 
+before(:each). chỉ thao tác trong mỗi khối it. sang khôi if khác khởi tạo lại từ đầu
+before(:all).  khi dùng muốn sử dụng những thay đổi các khối if khác nhau. khởi tạo một lần sử dụng trong tất cả các khối if

Reply:
mặc định là before each hay all e

Reply:
mặc định là before each . khi dùng chỉ cần viết before là nó tự hiểu là before(:each)